### PR TITLE
Fix overriding properties from parent context class

### PR DESCRIPTION
**Pull request by kpajdzik**

### CODE REVIEW

File: src/azure/Model/CodeModelTSa.cs [hunk 1]
```diff
@@ -143,6 +143,16 @@ protected override void GenerateNodeSampleImports(TSBuilder builder)
             GenerateNodeSampleMsRestNodeAuthImport(builder);
             GenerateNodeSampleClientImport(builder);
         }
+
+        protected override bool ShouldGenerateProperty(string propertyName)
+        {
+            string[] baseClassProperties = {
```
Comment (daschult) [inline src/azure/Model/CodeModelTSa.cs]: Please rename this to `azureServiceClientProperties`.

File: src/vanilla/Model/CodeModelTS.cs [hunk 2]
```diff
@@ -812,5 +812,36 @@ public string GenerateReadmeMdBrowserSampleCode(string emptyLine)
 
             return builder.ToString();
         }
+
+        public string GenerateClassFields(string emptyLine)
```
Comment (daschult) [inline src/vanilla/Model/CodeModelTS.cs]: `GenerateClassProperties`

File: src/vanilla/Model/CodeModelTS.cs [hunk 3]
```diff
@@ -812,5 +812,36 @@ public string GenerateReadmeMdBrowserSampleCode(string emptyLine)
...
+                builder.Property(property.Name, property.ModelType.TSType(false), property.IsRequired);
+            }
+
+            return builder.ToString();
+        }
+
+        protected virtual bool ShouldGenerateProperty(string propertyName)
+        {
+            string[] propertiesToIgnore = { "generateClientRequestId" };
+            string[] baseClassProperties = {
```
Comment (daschult) [inline src/vanilla/Model/CodeModelTS.cs]: `serviceClientProperties`

File: src/DSL/TSBuilder.cs [hunk 4]
```diff
@@ -633,6 +633,13 @@ public void ConstObjectVariable(string variableName, string value)
             ConstObjectVariable(variableName, null, value);
         }
 
+        public void Property(string name, string type, bool required, string accessModifier = "")
```
Comment (daschult) [inline src/DSL/TSBuilder.cs]: Please make `required` default to `true`.

File: src/vanilla/Model/CodeModelTS.cs [hunk 5]
```diff
@@ -812,5 +812,36 @@ public string GenerateReadmeMdBrowserSampleCode(string emptyLine)
...
         }
+
+        public string GenerateClassFields(string emptyLine)
+        {
+            TSBuilder builder = new TSBuilder();
+
+            foreach (Property property in Properties.Where(property => ShouldGenerateProperty(property.Name)))
+            {
+                builder.Line(emptyLine);
+                builder.Property(property.Name, property.ModelType.TSType(false), property.IsRequired);
```
Comment (daschult) [inline src/vanilla/Model/CodeModelTS.cs]: To keep prior behavior, please don't provided a value for `required` here.
```suggestion
                builder.Property(property.Name, property.ModelType.TSType(false));
```

File: src/vanilla/Model/CodeModelTS.cs (hunk 3 above)
Comment (sergey-shandar) [inline src/vanilla/Model/CodeModelTS.cs]: these variables should be `static readonly `

File: test/multiapi/packages/multiapi-test/.npmignore [hunk 6]
```diff
@@ -0,0 +1,35 @@
+#git
```
Comment (sergey-shandar) [inline test/multiapi/packages/multiapi-test/.npmignore]: This file should be removed.

File: src/azure/Templates/AzureServiceClientContextTemplate.cshtml [hunk 7]
```diff
@@ -24,11 +23,7 @@ const packageVersion = "@Model.Settings.PackageVersion";
 @{var optionalParameters = Model.Properties.Where(p => (!p.IsRequired || p.IsRequired && !string.IsNullOrEmpty(p.DefaultValue)) && !p.IsConstant && !p.Name.EqualsIgnoreCase("apiVersion"));}
 @EmptyLine
 export class @(Model.ContextName) extends msRestAzure.AzureServiceClient {
-@foreach (var property in Model.Properties.Where(p => p.Name != "generateClientRequestId"))
-{
-  @EmptyLine
-  @:@(property.Name): @(property.ModelType.TSType(false));
-}
+  @(Model.GenerateClassFields(EmptyLine))
```
Comment (sergey-shandar) [inline src/azure/Templates/AzureServiceClientContextTemplate.cshtml]: I like it. IMHO, we should reduce amount of C# code in `.cshtml` files.

File: test/multiapi/packages/multiapi-test/.npmignore (hunk 6 above)
Comment (kpajdzik) [reply test/multiapi/packages/multiapi-test/.npmignore]: This is autogenerated.

Comment (sergey-shandar) [reply test/multiapi/packages/multiapi-test/.npmignore]: Oh.. it shouldn't generate it after this PR https://github.com/Azure/autorest.typescript/pull/344. Could you try to delete the file and regenerate? If it's still generated, probably I missed something in my PR.

File: test/multiapi/packages/multiapi-test-2017-10-01/.npmignore [hunk 8]
```diff
@@ -0,0 +1,35 @@
+#git
```
Comment (sergey-shandar) [inline test/multiapi/packages/multiapi-test-2017-10-01/.npmignore]: another .npmignore?

File: src/vanilla/Model/CodeModelTS.cs [hunk 9]
```diff
@@ -812,5 +826,23 @@ public string GenerateReadmeMdBrowserSampleCode(string emptyLine)
...
             return builder.ToString();
         }
+
+        public string GenerateClassProperties(string emptyLine)
+        {
+            TSBuilder builder = new TSBuilder();
+
+            foreach (Property property in Properties.Where(property => ShouldGenerateProperty(property.Name)))
+            {
+                builder.Line(emptyLine);
```
Comment (daschult) [inline src/vanilla/Model/CodeModelTS.cs]: Remove this!!!!!!!

File: src/vanilla/Model/CodeModelTS.cs [hunk 10]
```diff
@@ -812,5 +826,23 @@ public string GenerateReadmeMdBrowserSampleCode(string emptyLine)
...
+                builder.Property(property.Name, property.ModelType.TSType(false));
```
Comment (daschult) [inline src/vanilla/Model/CodeModelTS.cs]: I'm dumb. Please add the required parameter back. :(

**Pull request merged**